Yosemite/Yosemite/Stores/ProductStore.swift

```diff
@@ -319,6 +321,35 @@ private extension ProductStore {
         }
     }
 
+    func retrieveProductsInCacheSortedByDescendingPopularity(siteID: Int64, onCompletion: @escaping ([Product]) -> Void) {
+        // Get completed orders
+        let completedOrderPredicate = NSPredicate(format: "statusKey ==[c] %@", OrderStatusEnum.completed.rawValue)
+        let sitePredicate = NSPredicate(format: "siteID == %lld", siteID)
+        let predicate = NSCompoundPredicate(andPredicateWithSubpredicates: [completedOrderPredicate, sitePredicate])
+        let completedStorageOrders = sharedDerivedStorage.allObjects(ofType: StorageOrder.self,
+                                                      matching: predicate,
+                                                      sortedBy: nil)
+
+        let completedOrders = completedStorageOrders.map { $0.toReadOnly() }
+
+        // Get product ids sorted by occurence
+        let completedOrdersItems = completedOrders.flatMap { $0.items }
+        let productIDCountDictionary = completedOrdersItems.reduce(into: [:]) { counts, orderItem in counts[orderItem.productID, default: 0] += 1 }
+        let sortedByOccurenceProductIDs = productIDCountDictionary.sorted { $0.value > $1.value }.map { $0.key }
+
+        // Retrieve products from product ids and finish
+        let products = sortedByOccurenceProductIDs
+            .compactMap {
+            let predicate = NSPredicate(format: "productID == %lld", $0)
+            let product = sharedDerivedStorage.allObjects(ofType: StorageProduct.self,
+                                                          matching: predicate,
+                                                          sortedBy: nil).first
```

Comment:
Following-up with making the action more generic, we could extract part of this method into a different function where the sorting happens based on a `sortOrder` parameter if we pass any sort of order, otherwise we just return all cached products in storage.

Reply:
Same as the previous question :)